dhis-2/dhis-services/dhis-service-tracker/src/main/java/org/hisp/dhis/tracker/export/event/JdbcPredicate.java

```diff
@@ -0,0 +1,230 @@
+package org.hisp.dhis.tracker.export.event;
+
+import static org.hisp.dhis.common.QueryFilter.affixLikeWildcards;
+import static org.hisp.dhis.system.util.SqlUtils.escapeLikeWildcards;
+import static org.hisp.dhis.system.util.SqlUtils.lower;
+import static org.hisp.dhis.system.util.SqlUtils.quote;
+
+import java.sql.Types;
+import java.util.List;
+import java.util.Optional;
+import javax.annotation.Nonnull;
+import org.hisp.dhis.common.QueryFilter;
+import org.hisp.dhis.common.QueryOperator;
+import org.hisp.dhis.common.UID;
+import org.hisp.dhis.common.ValueType;
+import org.hisp.dhis.common.ValueType.SqlType;
+import org.hisp.dhis.common.ValueTypedDimensionalItemObject;
+import org.hisp.dhis.dataelement.DataElement;
+import org.hisp.dhis.system.util.SqlUtils;
+import org.hisp.dhis.trackedentity.TrackedEntityAttribute;
+import org.springframework.jdbc.core.SqlParameterValue;
+import org.springframework.jdbc.core.namedparam.MapSqlParameterSource;
+
+/**
+ * JdbcPredicate turns a {@link QueryFilter} on a data element or tracked entity attribute into a
+ * SQL predicate with an optional JDBC parameter (right-hand side operand of the SQL operator). The
+ * {@link #sql()} can be inserted into a SQL statement where clause. User input in the {@link
+ * QueryFilter#getFilter()} is transformed into the {@link #parameter()}. Non-unary operators will
+ * have a parameter which must be added to the JDBC template parameter source like {@link
+ * MapSqlParameterSource}!
+ *
+ * <p>These are some of the transformations we need to do on user input
+ *
+ * <ul>
+ *   <li>split input {@code "one;two"} for operator {@link QueryOperator#IN}
+ *   <li>convert input value(s) from text to other data types so comparisons can be done using DHIS2
+ *       data value type semantics i.e. {@code 170.0 == 170} which is what users expect for value
+ *       type integer/number while {@code "170.0" != "170"} is unexpected but what happens for
+ *       treating the values as text
+ *   <li>escape wildcard characters '%' and '_' for operators backed by SQL {@code like}
+ *   <li>insert '%' wildcard(s) before/after user input for like based operators
+ * </ul>
+ *
+ * Other special characters like single quotes are expected to be escaped by the <a
+ * href="https://github.com/pgjdbc/pgjdbc/blob/156d724e1d95052b41a19fb568b2f81919ae2197/pgjdbc/src/main/java/org/postgresql/core/v3/SimpleParameterList.java#L213">JDBC
+ * driver</a>!
+ *
+ * <p>On {@link SqlParameterValue}: Spring handles {@code List.of()} values for us no matter if we
+ * pass it to an <a
+ * href="https://docs.spring.io/spring-framework/reference/data-access/jdbc/parameter-handling.html#jdbc-in-clause">in</a>
+ * operator allowing multiple values or an operator only expecting one like {@code eq}.
+ * JdbcPredicate ensures we never pass a collection with multiple elements to such operators as this
+ * leads to a {@code BadSqlGrammarException}.
+ */
+record JdbcPredicate(String sql, Optional<Parameter> parameter) {
```

Comment:
Is there a way to make the constructor of the record private?
I understand that we should create `JdbcPredicate`s only with the static constructors, so it would be good to enforce that at code level.

Reply:
We could only do that by making it a class with a private constructor. Should we?

Reply:
Maybe.
After all your work if someone arrives tomorrow and creates a `new JdbcPredicate` then all the consistency is gone.